client.go

```diff
@@ -40,35 +40,69 @@ func init() {
 
 // StatsdClient is a client library to send events to StatsD
 type StatsdClient struct {
-	conn             net.Conn
-	addr             string
-	prefix           string
-	eventStringTpl   string
-	Logger           Logger
-	conn_type        string
-	reconnect_ticker *time.Ticker
+	conn            net.Conn
+	addr            string
+	prefix          string
+	eventStringTpl  string
+	Logger          Logger
+	connType        string
+	reconnectTicker *time.Ticker
 }
 
-// NewStatsdClient - Factory
-func NewStatsdClient(addr string, prefix string) *StatsdClient {
+// ConfigurationFunc is a typedef for a function that configures some aspect of
+// a StatsdClient instance
+type ConfigurationFunc func(*StatsdClient)
+
+// AutoReconnect returns a ConfigurationFunc that causes the StatsdClient to automatically
+// recreate its underlying connection on the specified interval.
+func AutoReconnect(interval time.Duration) ConfigurationFunc {
+	return func(client *StatsdClient) {
+		if interval > 0 {
+			client.reconnectTicker = time.NewTicker(interval)
+		}
+	}
+}
+
+// UseLogger returns a ConfigurationFunc that makes the StatsdClient use the specified Logger
+// implementation, rather than the default implementation.
+func UseLogger(logger Logger) ConfigurationFunc {
+	return func(client *StatsdClient) {
+		client.Logger = logger
+	}
+}
+
+// NewStatsdClient is a factory func that creates a StatsdClient that sends to
+// the configured address and prefixes all stats with the given prefix name.
+func NewStatsdClient(addr string, prefix string, options ...ConfigurationFunc) *StatsdClient {
 	// allow %HOST% in the prefix string
 	prefix = strings.Replace(prefix, "%HOST%", Hostname, 1)
 	client := &StatsdClient{
-		addr:             addr,
-		prefix:           prefix,
-		Logger:           log.New(os.Stdout, "[StatsdClient] ", log.Ldate|log.Ltime),
-		eventStringTpl:   "%s%s:%s",
-		reconnect_ticker: time.NewTicker(30 * time.Second),
```

Comment:
@dcarney do we want a default interval?   Depending on the implementation, of course.  But I'd imagine we'd alway want the ability to reconnect if there's a failure.

Reply:
@dcarney This seems to be another case of a flaw in the functional option pattern. Is there an elegant way around this?

Reply:
@Jonnymcc The functional options pattern doesn't preclude you from having defaults. In fact, if no "custom" logger is specified using a functional option, we provide a default implementation:

https://github.com/HBOCodeLabs/statsd/blob/dcarney/reconnect-logging/client.go#L90-L93

I was thinking that the re-connect behavior would be opt-in only, but if you guys feel like it should be the default, I can provide a default reconnection interval.

Reply:
@Jonnymcc @jsongHBO I just changed this to have a default reconnection interval.  PTAL.